src/rai_bench/rai_bench/tool_calling_agent/mocked_tools.py

```diff
@@ -265,45 +268,177 @@ def _run(self, msg_type: str) -> str:
+
+        Parameters
+        ----------
+        service_type : str
+            The ROS 2 service type in format "package_name/srv/ServiceName".
+
+        Returns
+        -------
+        class
+            The dynamically imported ROS 2 service class.
+
+        Raises
+        ------
+        ValueError
+            If the service_type format is invalid (not in format
+            "package_name/srv/ServiceName").
+
+        Notes
+        -----
+        Results are cached in ros2_services_cache to avoid repeated imports
+        of the same service type.
+        """
+        if service_type in self.ros2_services_cache:
+            return self.ros2_services_cache[service_type]
+
+        # Parse service type: "package_name/srv/ServiceName"
+        parts = service_type.split("/")
+        if len(parts) != 3 or parts[1] != "srv":
+            raise ValueError(f"Service type: {service_type} is invalid")
+
+        package_name, _, service_name = parts
+
+        module = importlib.import_module(f"{package_name}.srv")
+        service_class = getattr(module, service_name)
+
+        self.ros2_services_cache[service_type] = service_class
+        return service_class
+
+    def validate_with_ros2(self, service_type: str, args: Dict[str, Any]):
+        """Validate using installed ROS2 packages services definition
+
+        Parameters
+        ----------
+        service_type : str
+        args : Dict[str, Any]
+            Dictionary of arguments to validate against the service definition.
+
+        Raises
+        ------
+        TypeError
+            When service type does not exist in ROS2 installed packages
+        """
+        service_class = self.get_ros2_service_class(service_type)
+        if not service_class:
+            raise TypeError(f"Service type: {service_type} does not exist.")
+
+        request = service_class.Request()
+        # set message fields converts them to object so we need deepcopy to avoid it
+        args_to_validate = copy.deepcopy(args)
+        set_message_fields(request, args_to_validate)
+
+    def validate_with_custom(self, service_type: str, args: Dict[str, Any]):
+        """
+        Validate using Pydantic model of custom messages.
+
+        Parameters
+        ----------
+        service_type : str
+        args : Dict[str, Any]
+            Dictionary of arguments to validate against the Pydantic model.
+
+        Raises
+        ------
+        ValueError
+            If service_type is not found in custom_models or if Pydantic
+            validation fails.
+        """
+        if service_type not in self.custom_models:
+            raise ValueError(f"Service type: {service_type} is invalid custom type")
+
+        model = self.custom_models[service_type]
+        try:
+            model.model_validate(args)
+        except ValidationError as e:
+            raise ValueError(f"Pydantic validation failed: {e}") from e
+
+    def validate(self, service_type: str, args: Dict[str, Any]):
```

Comment:
what happens when the validation fails?
In case of `validate_with_custom` a ValueError is risen. 
How about `validate_with_ros2`?